_posts/2018-08-30-on-the-fly-transformations.md

```diff
@@ -0,0 +1,283 @@
+Fitting is useful when processing trajectories for visualization and analyses - it removes the translations and rotations of the molecule, allowing us to have a better look at the structural changes that happen in our simulations.
+If we want to do this using `trjconv` we would do have to do this in two steps:
+
+     gmx trjconv -f pept_in_memb.xtc -s pept_in_memb.tpr -pbc mol -center -o midstep.xtc
+    
+     gmx trjconv -f midstep.xtc -s pept_in_memb.tpr -fit rot+trans -o output.xtc
+
+And we choose `Protein` as the group to be centered and for the least squares fitting.
+
+In MDAnalysis we just add another transformation to our workflow - `fit_rot_trans`. This transformation takes the AtomGroup to be fitted as argument, an AtomGroup to be used as reference and, by default, it behaves just as the option `-fit rot+trans`. If given a `plane` argument, the fitting is performed on a given plane. If `plane=xy` then the transformation will behave as `-fit rotxy+transxy`, but the `xz` and `yz` planes are also supported. Just as in `center_in_box`, a `weights` argument can be passed to the function, and it will dictate how much each atom of the molecule contributes to the least squares fitting. 
+Here's what the workflow looks like:
+
+
+```python
+u = mda.Universe('pept_in_memb.tpr', 'pept_in_memb.xtc')
+prot = u.select_atoms("protein")
+# we load another universe to define the reference
+# it uses the same input files, but this doesn't have to be always the case
+ref_u = mda.Universe('pept_in_memb.tpr', 'pept_in_memb.xtc')
+reference = ref_u.select_atoms("protein")
+ag = u.atoms
+transformations = (mda.transformations.unwrap(ag),
+                   mda.transformations.center_in_box(prot, weights='mass'),
+                   mda.transformations.wrap(ag, compound='fragments'),
+                   mda.transformations.fit_rot_trans(prot, reference))
+u.trajectory.add_transformations(*transformations)
+nv.show_mdanalysis(u)
+
+```
+
+
+    NGLWidget(count=11)
+
+
+It looks a bit confusing with the membrane...
+
+
+```python
+t = nv.MDAnalysisTrajectory(prot)
+w = nv.NGLWidget(t)
+w.add_line()
+w
+```
+
+
+    NGLWidget(count=11)
+
+
+This transformations is good when we want to see how the conformation of the protein evolves with time. 
+
+But, in this case, we also have a membrane. How does the protein behave in the membrane? Doing a least squares fitting in the `xy` plane can help us have a better look. Here's how it goes:
+
+
+```python
+u = mda.Universe('pept_in_memb.tpr', 'pept_in_memb.xtc')
+prot = u.select_atoms("protein")
+ref_u = mda.Universe('pept_in_memb.tpr', 'pept_in_memb.xtc')
+reference = ref_u.select_atoms("protein")
+ag = u.atoms
+transformations = (mda.transformations.unwrap(ag),
+                   mda.transformations.center_in_box(prot),
+                   mda.transformations.wrap(ag, compound='fragments'),
+                   mda.transformations.fit_rot_trans(prot, reference, plane='xy', weights="mass"))
```

Comment:
Here you can enter another wrap call. This means, in the end, all atoms are nicely within the unit-cell without any lipids wandering off. That is especially a transformation that is not possible with gromacs tools.

Reply:
Wrapping after a rotation is not advised. The rotation breaks the relationship between the coordinate and the box so the wrapping will be erroneous.

There is a comment about this in at least one of the docstring. Note also that the issue is not specific to MDAnalysis but that gromacs has it as well. 

There is a solution for that but it is not trivial. The box matrix could be rotated, but it is not a common thing to do and it will likely break assumptions about how triclinic boxes are handled both in MDAnalysis and other softwares.

Reply:
Unrelated to the PR, but related to this topic: is it impossible to center the system on a protein, i.e. helix lying on the membrane, and wrap the lipids such that the membrane does not rotate in circles? This would be neat for visualization / density calculations of membrane parameters in relation to the protein.

As far as I'm aware it's impossible to do with GROMACS, but I was hoping that the related PR would enable the functionality in MDAnalysis.

Reply:
Do you mean fit instead of center? Because if you just center the helix in the unit cell then there are no rotation operations performed on the coordinates. Thus you can still perform pbc transformations - wrap and unwrap. This can also be done in gromacs.
If you do rotational fitting though, there are no transformations currently in MDAnalysis that can do that. As you rotate the system, it becomes "disconnected" from the unit cell , making pbc transformations trickier. It can be done, but the resulting system would be different - in membrane systems it would be mostly fine if the membrane is pure. But for mixtures it would throw the lipid disposition off.

Reply:
Possibly add a footnote to discuss this case.

I agree with @mimischi that I'd love to have the unwrap after rotation functionality. Maybe another GSoC project...